activerecord/test/cases/store_test.rb

```diff
@@ -191,4 +191,42 @@ class StoreTest < ActiveRecord::TestCase
     second_dump = YAML.dump(loaded)
     assert_equal @john, YAML.load(second_dump)
   end
+
+  test "setting type casted values on build" do
+    jamie = Admin::User.new(name: "Jamie", active: "true", salary: 3.1999, birthday: '2000-01-01')
+    assert_equal true, jamie.active
+    assert_equal 3, jamie.salary
+    assert_equal Date.new(2000,1,1), jamie.birthday
+  end
+
+  test "setting type casted values from db" do
+    jamie = Admin::User.new(name: "Jamie")
+    jamie.active = "t"
+    jamie.salary = 3.1999 
+    jamie.birthday = '01/01/2000'
+    jamie.save!
+
+    jamie = Admin::User.where(name: "Jamie").first
+   
+    assert_equal '2000-01-01', jamie.json_data['birthday']    
```

Comment:
The simplest solution would be to have the store accessors create a new hash, rather than mutate the existing one. Of course this is fragile for obvious reasons. We would also still have to consider the case of `store['date'] = '2015-02-25'` not type casting. Agreed that we should not return `TypedHash` or do anything of that nature.

While I know the case you're trying to handle is fairly simple, please understand we don't want to handle the simple case when it leads to unexpected behavior in a ton of edge cases, as this tends to be harder to fix later.

I think there's a solution here, but the way we go about implementing `store` might need to be rethought. Let me think about it for a bit and get back to you.